File: cc/058.cc
// Project Euler Problem 058
//
// Investigate the number of primes that lie on the diagonals of the
// spiral grid.

#include <cstdio>
#include "primes1.h"

int main(int argc, const char* argv[]) {
  Primes1<long int> primes;
  int n = 1;
  int p = 0;
  int last = 1;
  int side = 0;
  while (!p || (p*10 > n)) {
    side += 2;
    for (int i=0; i<4; i++) {
      last += side;
      n++;
      if (primes.lazy_is_prime(last)) p++;
    }
  }
  printf("%d\n", side+1);
}

File: go/problems/058.go
// Project Euler Problem 058
//
// Investigate the number of primes that lie on the diagonals of the
// spiral grid.

package problems

import (
	"fmt"

	"../primes"
)

func Problem058() string {
	last, length, prime, total := 9, 3, 3, 5
	for prime*10 > total {
		length += 2
		for i := 0; i < 4; i++ {
			last += length - 1
			total++
			if primes.IsPrime(last) {
				prime++
			}
		}
	}
	return fmt.Sprint(length)
}

func init() {
	Functions[58] = Problem058
}
